roles/kubernetes-apps/ansible/tasks/cleanup_dns.yml

```diff
@@ -0,0 +1,54 @@
+---
+- name: Kubernetes Apps | Delete old CoreDNS resources
+  kube:
+    name: "coredns"
+    namespace: "{{ system_namespace }}"
+    kubectl: "{{ bin_dir }}/kubectl"
+    resource: "{{ item }}"
+    state: absent
+  with_items:
+    - 'deploy'
+    - 'configmap'
+    - 'svc'
+  tags:
+    - upgrade
+
+- name: Kubernetes Apps | Delete kubeadm CoreDNS
+  kube:
+    name: "coredns"
+    namespace: "{{ system_namespace }}"
+    kubectl: "{{ bin_dir }}/kubectl"
+    resource: "deploy"
+    state: absent
+  when:
+    - kubeadm_enabled|default(false)
+    - kubeadm_init.changed|default(false)
+    - inventory_hostname == groups['kube-master'][0]
+
+- name: Kubernetes Apps | Delete old KubeDNS resources
```

Comment:
This should only happen when necessary. Right now it will delete kubedns always on every run, which is not what we want. So you should add a when clause to delete his if the mode is not kubedns or dnsmasq_kubedns

Reply:
This is the current behavior of kubespray. See [here](https://github.com/kubernetes-incubator/kubespray/blob/master/roles/kubernetes-apps/ansible/tasks/main.yml#L27) . So impemented the same behavior in this PR.

Reply:
But that cleanup was done only if kubeadm was the deploy mode.